File: src/test/java/stepDefinitions/AccountSteps.java
package stepDefinitions;

import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;
import pages.BaseTest;
import pages.DashboardPage;
import pages.LoginPage;
import pages.SignupPage;
import utils.DataGenerator;

public class AccountSteps {

    SignupPage signupPage;
    LoginPage loginPage;
    DashboardPage dashboardPage;

    String username = DataGenerator.generateUsername();
    String password = DataGenerator.generatePassword();

    @Given("the user is on the signup page")
    public void user_on_signup_page() {
        signupPage = new SignupPage(BaseTest.page);
        signupPage.navigate();
    }

    @When("the user registers with valid details")
    public void user_registers() {
        System.out.println("Registering: " + username + " / " + password);
        signupPage.register("John", "Doe", username, password);


         signupPage.logOut();
    }

    @When("logs in with the same credentials")
    public void logs_in() {
        loginPage = new LoginPage(BaseTest.page); // fixed page reference
        loginPage.navigate();
        System.out.println("Logging in: " + username + " / " + password);
        loginPage.login(username, password);
    }

    @Then("the user should see the account balance displayed")
    public void verify_balance() {
        dashboardPage = new DashboardPage(BaseTest.page);
        String balance = dashboardPage.getBalance();
        System.out.println("Account Balance: " + balance);

        BaseTest.takeScreenshot("balance.png");
    }
}

File: src/steps/account.steps.ts
import { Given, Then, When } from '@cucumber/cucumber';
import { DashboardPage } from '../pages/dashboard.page';
import { LoginPage } from '../pages/login.page';
import { SignupPage } from '../pages/signup.page';
import { page, takeScreenshot } from '../support/test-base';
import { generatePassword, generateUsername } from '../utils/data-generator';

let signupPage: SignupPage;
let loginPage: LoginPage;
let dashboardPage: DashboardPage;

const username = generateUsername();
const password = generatePassword();

Given('the user is on the signup page', async function () {
  signupPage = new SignupPage(page);
  await signupPage.navigate();
});

When('the user registers with valid details', async function () {
  console.log(`Registering: ${username} / ${password}`);
  await signupPage.register(username, password);
  await signupPage.logOut();
});

When('logs in with the same credentials', async function () {
  loginPage = new LoginPage(page);
  await loginPage.navigate();
  console.log(`Logging in: ${username} / ${password}`);
  await loginPage.login(username, password);
});

Then('the user should see the account balance displayed', async function () {
  dashboardPage = new DashboardPage(page);
  const balance = await dashboardPage.getBalance();
  console.log(`Account Balance: ${balance}`);
  await takeScreenshot('balance.png');
});
